assignment1/src/main/java/ro/utcn/sd/he/assignment1/persistence/jpa/HibernateQuestionTagRepository.java

```diff
@@ -0,0 +1,64 @@
+package ro.utcn.sd.he.assignment1.persistence.jpa;
+
+import lombok.RequiredArgsConstructor;
+import ro.utcn.sd.he.assignment1.model.Question;
+import ro.utcn.sd.he.assignment1.model.QuestionTag;
+import ro.utcn.sd.he.assignment1.model.Tag;
+import ro.utcn.sd.he.assignment1.model.User;
+import ro.utcn.sd.he.assignment1.persistence.api.QuestionTagRepository;
+import ro.utcn.sd.he.assignment1.service.QuestionTagService;
+
+import javax.persistence.EntityManager;
+import javax.persistence.criteria.CriteriaBuilder;
+import javax.persistence.criteria.CriteriaQuery;
+import javax.persistence.criteria.Root;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+
+@RequiredArgsConstructor
+public class HibernateQuestionTagRepository implements QuestionTagRepository {
+    private final EntityManager entityManager;
+
+    @Override
+    public void insert(Question question, Tag tag) {
+        QuestionTag questionTag = new QuestionTag(question.getId(),tag.getId());
+        entityManager.persist(questionTag);
+    }
+
+    @Override
+    public void remove(Question question, Tag tag) {
+        QuestionTag questionTag = new QuestionTag(question.getId(),tag.getId());
+        entityManager.remove(questionTag);
+    }
+
+    @Override
+    public List<Question> getQuestionsWithTag(Tag tag) {
+        CriteriaBuilder criteriaBuilder = entityManager.getCriteriaBuilder();
+        CriteriaQuery criteriaQuery = criteriaBuilder.createQuery(QuestionTag.class);
+        Root<QuestionTag> questionRoot = criteriaQuery.from(QuestionTag.class);
+        List<QuestionTag> questionTags = entityManager.createQuery(criteriaQuery.select(questionRoot).where(criteriaBuilder.equal(questionRoot.get("tagID"),tag.getId()))).getResultList();
+        //since i dod not implement the many to many relationship i have to do this trick where I get this ^ list which basically contains all the question Ids that i need
+        HibernateQuestionRepository questionRepository = new HibernateQuestionRepository(entityManager);
+        List<Question> questions = new ArrayList<>();
+        for(QuestionTag questionTag : questionTags){
```

Comment:
When you do a for-loop and inside you run an SQL query, it is a BIG code smell :) For example in this case, you could simply have another repository method (just for the Hibernate repo) `findByIdsIn(List<Integer> ids)` and do a single query using the [IN](https://stackoverflow.com/a/13145364/7612556) predicate.

Reply:
That is a great idea. Thank you very much for all the feedback, it's really helpful.